number_theory/modular.h

```diff
@@ -0,0 +1,82 @@
+#ifndef NUMBER_THEORY_MODULAR_H_
+#define NUMBER_THEORY_MODULAR_H_
+
+#include <stddef.h>
+
+#include <bit>
+#include <limits>
+
+#include "number_theory/utility.h"
+
+// Modular arithmetic
+
+namespace tql {
+namespace number_theory {
+
+template <class T, T mod>
+class Modular {
+  static_assert(std::numeric_limits<T>::is_integer && mod > 0,
+                "Modular requires modulus to be a positive integer.");
+
+ public:
+  using type = T;
+  static constexpr T modulus = mod;
+
+  Modular(type value = 0) { set(std::move(value)); }
+  Modular(const Modular &other) = default;
+  Modular(Modular &&other) = default;
+
+  Modular &operator=(const Modular &other) = default;
+  Modular &operator=(Modular &&other) = default;
+
+  const type &get() const { return value_; }
+
+  void set(type value) {
+    value_ = value % modulus;
+    if (value_ < 0)
+      value_ += modulus;
+  }
+
+  Modular add(const Modular &rhs) const {
+    check_addition_overflow();
+    return Modular((value_ + rhs.value_) % modulus);
+  }
+
+  Modular negate() const { return Modular(modulus - value_); }
+
+  Modular substract(const Modular &rhs) const { return add(rhs.negate()); }
+
+  Modular multiply(const Modular &rhs) const {
+    check_multiplication_overflow();
+    return Modular((value_ * rhs.value_) % modulus);
```

Comment:
What about using binary exponentiation trick when we know that it can overflow?

https://cp-algorithms.com/algebra/binary-exp.html#variation-of-binary-exponentiation-multiplying-two-numbers-modulo-m

Reply:
I plan to implement this trick in a future PR.